### PR TITLE
[Storage Blobs Service] Blob Versioning

### PR DESCRIPTION
**Pull request by JinmingHu-MSFT**

### CODE REVIEW

File: sdk/storage/src/blobs/blob_sas_builder.cpp [hunk 1]
```diff
@@ -120,8 +121,9 @@ namespace Azure { namespace Storage { namespace Blobs {
     std::string stringToSign = Permissions + "\n" + (StartsOn.HasValue() ? StartsOn.GetValue() : "")
         + "\n" + ExpiresOn + "\n" + canonicalName + "\n" + Identifier + "\n"
         + (IPRange.HasValue() ? IPRange.GetValue() : "") + "\n" + protocol + "\n" + Version + "\n"
-        + resource + "\n" + Snapshot + "\n" + CacheControl + "\n" + ContentDisposition + "\n"
-        + ContentEncoding + "\n" + ContentLanguage + "\n" + ContentType;
+        + resource + "\n" + (!Snapshot.empty() ? Snapshot : BlobVersionId) + "\n" + CacheControl
```
Comment (JinmingHu-MSFT) [inline sdk/storage/src/blobs/blob_sas_builder.cpp:124]: maybe I should choose between snapshot and versionid according to resource type.

File: sdk/storage/test/blobs/block_blob_client_test.cpp [hunk 2]
```diff
@@ -163,23 +166,36 @@ namespace Azure { namespace Storage { namespace Test {
...
     EXPECT_EQ(snapshotClient.GetProperties()->Metadata, m_blobUploadOptions.Metadata);
+    auto versionClient = m_blockBlobClient->WithVersionId(res->VersionId.GetValue());
+    EXPECT_EQ(ReadBodyStream(versionClient.Download()->BodyStream), m_blobContent);
+    EXPECT_EQ(versionClient.GetProperties()->Metadata, m_blobUploadOptions.Metadata);
     auto emptyContent = Azure::Core::Http::MemoryBodyStream(nullptr, 0);
     EXPECT_THROW(snapshotClient.Upload(&emptyContent), StorageError);
     EXPECT_THROW(snapshotClient.SetMetadata({}), StorageError);
-    EXPECT_THROW(
-        snapshotClient.SetAccessTier(Azure::Storage::Blobs::AccessTier::Cool), StorageError);
+    EXPECT_NO_THROW(snapshotClient.SetAccessTier(Azure::Storage::Blobs::AccessTier::Cool));
```
Comment (JinmingHu-MSFT) [inline sdk/storage/test/blobs/block_blob_client_test.cpp:180]: This feature is not enabled for some storage accounts.

Comment (JinmingHu-MSFT) [reply sdk/storage/test/blobs/block_blob_client_test.cpp:180]: https://github.com/Azure/azure-sdk-for-cpp/issues/415

File: sdk/storage/inc/blobs/page_blob_client.hpp [hunk 3]
```diff
@@ -98,6 +98,16 @@ namespace Azure { namespace Storage { namespace Blobs {
...
 
+    /**
+     * @brief Creates a clone of this instance that references a version ID rather than the base
+     * blob.
+     *
+     * @param versionId The version ID returning a URL to the base blob.
+     * @return A new PageBlobClient instance.
+     * @remarks Pass empty string to remove the version ID returning the base blob.
+     */
+    PageBlobClient WithVersionId(const std::string& versionId) const;
```
Comment (katmsft) [inline sdk/storage/inc/blobs/page_blob_client.hpp:109]: `VersionID`? Two character acronym should be all caps.
ditto in other places.

Comment (JinmingHu-MSFT) [reply sdk/storage/inc/blobs/page_blob_client.hpp:109]: This is not an acronym

File: sdk/storage/src/blobs/blob_client.cpp [hunk 4]
```diff
@@ -135,6 +135,20 @@ namespace Azure { namespace Storage { namespace Blobs {
...
+  BlobClient BlobClient::WithVersionId(const std::string& versionId) const
+  {
+    BlobClient newClient(*this);
+    if (versionId.empty())
+    {
+      newClient.m_blobUrl.RemoveQuery(Details::c_HttpQueryVersionId);
+    }
+    else
+    {
+      newClient.m_blobUrl.AppendQuery(Details::c_HttpQueryVersionId, versionId);
```
[the remark this answers is not in the document]
Comment (JinmingHu-MSFT) [reply sdk/storage/src/blobs/blob_client.cpp:147]: it'll replace the old value

File: sdk/storage/inc/blobs/page_blob_client.hpp (hunk 3 above)
Comment (katmsft) [reply sdk/storage/inc/blobs/page_blob_client.hpp:109]: It's an abbreviation. @BillyONeal Do rules apply to acronym also apply to abbreviation?

Comment (JinmingHu-MSFT) [reply sdk/storage/inc/blobs/page_blob_client.hpp:109]: it says `Acronyms and initialisms`. https://azure.github.io/azure-sdk/cpp_design.html#definitions

**Pull request merged**